lib/puppet/type/user.rb

```diff
@@ -171,10 +171,15 @@ def sync
 
     newproperty(:comment) do
       desc "A description of the user.  Generally the user's full name."
-      if RUBY_VERSION < "2.1.0"
-        munge do |v|
-          v.respond_to?(:force_encoding) ? v.force_encoding(Encoding::ASCII_8BIT) : v
+      def insync?(current)
+        # nameservice provider requires special attention to encoding
+        # Use Puppet::Property#should instead of @should to avoid extra
+        # array handling logic
+        if !@should.empty? and provider.respond_to?(:comments_insync?)
```

Comment:
Great question about whether we want the `munged` vs raw value (or even also `unmunged`). Turns out we call various munging and unmunging methods throughout the lifecycle of the property, but at the time of comparing is/should values for synced-ness, they are:
- is-value: a value that has been unmunged by the `unmunge` method defined, if any, on the corresponding _option_ in the provider ([example](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/user/useradd.rb#L35-L42)). This is true at least for providers inheriting from nameservice.
- should-value: a value that has been munged by the `munge` method defined, if any, on the property in the type definition ([example, just imagine unmunge instead of munge](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/user.rb#L119-L128))

So what we want for our should when overriding is not the unmunged value returned by `Puppet::Property#should` but rather the munged value in the `@should` array. That being said, none of the comment options define munge or unmunge and the type definition only defines munge (and won't anymore, after this PR), but they might some day down the road so it's good to get this right. Thanks for the pointing it out, will update.